semanticdb/metacp/src/main/scala/scala/meta/cli/Metacp.scala

```diff
@@ -1,25 +1,94 @@
 package scala.meta.cli
 
 import java.io._
+import java.nio.file.Files
+
+import io.github.soc.directories.ProjectDirectories
+import org.langmeta.internal.io.PlatformFileIO
+import org.langmeta.io.AbsolutePath
+import org.langmeta.io.Classpath
+
 import scala.meta.internal.metacp._
 
 object Metacp {
   def main(args: Array[String]): Unit = {
     sys.exit(process(args, System.out, System.err))
   }
 
-  @deprecated("Use `process(args, System.out, System.err)` instead.", "3.4.0")
   def process(args: Array[String]): Int = {
     process(args, System.out, System.err)
   }
 
   def process(args: Array[String], out: PrintStream, err: PrintStream): Int = {
-    Settings.parse(args.toList) match {
+    Settings.parse(args.toList, out, err) match {
       case Some(settings) =>
-        val main = new Main(settings, out, err)
-        main.process()
+        val mclasspath = process(settings)
+        out.println(mclasspath.mkString(File.pathSeparator))
+        0
       case None =>
         1
     }
   }
+
+  def process(settings: Settings): List[AbsolutePath] = {
+    settings.classpath.shallow.toParArray.map(p => processPath(p, settings)).toList
+  }
+
+  def processPath(in: AbsolutePath, settings: Settings): AbsolutePath = {
+    if (in.isDirectory) {
+      val out = settings.directories match {
+        case DirectoryOutput.TempDirectory =>
+          AbsolutePath(Files.createTempDirectory("semanticdb"))
+        case DirectoryOutput.InPlace =>
+          in
+      }
+      val exit = Main.process(settings.copy(classpath = Classpath(in :: Nil), d = out))
+      assert(exit == 0, s"Failed to process $in")
+      out
+    } else if (!in.isFile) {
+      throw new IllegalArgumentException(s"$in is not a regular file")
+    } else {
+      val checksum = Checksum(in)
+      val out =
+        cacheFile(settings, in.toNIO.getFileName.toString.stripSuffix(".jar") + "-" + checksum)
+      if (!out.isFile) {
+        PlatformFileIO.withJarFileSystem(out) { jar =>
+          val exit = Main.process(settings.copy(classpath = Classpath(in :: Nil), d = jar))
+          assert(exit == 0, s"Failed to process $in")
+        }
+      }
+      out
+    }
+  }
+
+  def scalaLibrarySynthetics(scalaVersion: String, settings: Settings): AbsolutePath = {
```

Comment:
Do metacp clients need to invoke this method manually? Can we instead invoke it automatically whenever someone passes us scala-library.jar? This will further simplify the public API.

Reply:
One problem with relying on finding a classpath entry containing the string "scala-library" is that it won't work for

* fat jar classpaths
* scala-library compiled from source

By exposing this method we delegate to the client to deal with these cases instead of trying to provide an automatic solution that fails in some cases.

Reply:
I've been thinking about looking into `META-INF/MANIFEST.MF`, which must say:

```
15:00 ~/Downloads$ cat META-INF/MANIFEST.MF
Manifest-Version: 1.0
Bnd-LastModified: 1507705065521
Bundle-ManifestVersion: 2
Bundle-Name: Scala Standard Library
Bundle-RequiredExecutionEnvironment: JavaSE-1.8
Bundle-SymbolicName: org.scala-lang.scala-library
Bundle-Version: 2.12.4.v20171010-220511-VFINAL-c2a5883
Created-By: 1.8.0_121 (Oracle Corporation)
...
```

This probably won't work well with far jars though.

Reply:
If we don't do this automatically, how would generation of scala library synthetics work from the command line? Will `metap $(metacp $(coursier fetch scala-library))` include the synthetics?

Reply:
Let's add --include-scala-library-synthetics  to metacp to optionally include scala library synthetics in the output or not.

Reply:
Implemented in https://github.com/scalameta/scalameta/pull/1394.